jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionData.java

```diff
@@ -175,7 +175,14 @@ public SessionData(String id, String cpath, String vhost, long created, long acc
         _lastAccessed = lastAccessed;
         _maxInactiveMs = maxInactiveMs;
         calcAndSetExpiry();
-        _attributes = attributes;
+        setAttributes(attributes);
+    }
+
+    private void setAttributes(Map<String, Object> attributes) {
+        if(attributes != null) {
+            boolean isConcurrentHashMap = (attributes instanceof ConcurrentHashMap);
+            _attributes = isConcurrentHashMap ? attributes : new ConcurrentHashMap<>(attributes);
+        }
     }
```

Comment:
I think we're almost there! I think the best solution, which would not need a new method, is to simply reorganize the constructors like so:
```
     public SessionData(String id, String cpath, String vhost, long created, long accessed, long lastAccessed, long maxInactiveMs)
-    {
-        this(id, cpath, vhost, created, accessed, lastAccessed, maxInactiveMs, new ConcurrentHashMap<String, Object>());
-    }
-
-    public SessionData(String id, String cpath, String vhost, long created, long accessed, long lastAccessed, long maxInactiveMs, Map<String, Object> attributes)
     {
         _id = id;
         setContextPath(cpath);
@@ -176,7 +171,13 @@ public class SessionData implements Serializable
         _lastAccessed = lastAccessed;
         _maxInactiveMs = maxInactiveMs;
         calcAndSetExpiry();
-        _attributes = attributes;
+        _attributes = new ConcurrentHashMap<String, Object>();
+    }
+
+    public SessionData(String id, String cpath, String vhost, long created, long accessed, long lastAccessed, long maxInactiveMs, Map<String, Object> attributes)
+    {
+        this(id, cpath, vhost, created, accessed, lastAccessed, maxInactiveMs);
+        putAllAttributes(attributes);
     }
```

And sorry to be a pain, but the jetty coding style is to have braces on newlines.